File: Euclid/Game/Room/RoomManager.cs
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using Euclid.Storage.Database.Access;
using Euclid.Storage.Database.Data;
using Euclid.Util.Extensions;

namespace Euclid.Game
{
    public class RoomManager : ILoadable
    {
        #region Fields

        public static readonly RoomManager Instance = new RoomManager();

        #endregion

        #region Properties

        public ConcurrentDictionary<int, Room> Rooms { get; private set; }
        public List<RoomModel> RoomModels { get; private set; }

        #endregion

        #region Constructors

        public RoomManager()
        {
            Rooms = new ConcurrentDictionary<int, Room>();
        }

        public void Load()
        {
            RoomModels = RoomDao.GetModels().Select(x => new RoomModel(x)).ToList();
        }

        #endregion

        #region Public methods

        /// <summary>
        /// Get if the room is loaded
        /// </summary>
        public bool HasRoom(int roomId)
        {
            return Rooms.TryGetValue(roomId, out _);
        }

        /// <summary>
        /// Remove room from list
        /// </summary>
        public void RemoveRoom(int roomId)
        {
            Rooms.Remove(roomId);
        }

        /// <summary>
        /// Add room to the map of loaded rooms
        /// </summary>
        public void AddRoom(Room room)
        {
            if (room == null)
                return;

            if (Rooms.ContainsKey(room.Data.Id))
                return;

            Rooms.TryAdd(room.Data.Id, room);
        }

        /// <summary>
        /// Get room instance, else return newly created instance if room exists
        /// </summary>
        public Room GetRoom(int roomId)
        {

            if (Rooms.TryGetValue(roomId, out var room))
                return room;

            var data = RoomDao.GetRoomData(roomId);

            if (data != null) {
                return new Room(data);
            }

            return null;
        }

        /// <summary>
        /// Replace the room data retrieved from the database with actual room instances
        /// </summary>
        public List<Room> ReplaceQueryRooms(List<RoomData> roomsFromDatabase)
        {
            List<Room> rooms = new List<Room>();

            foreach (var roomData in roomsFromDatabase)
            {
                if (Rooms.TryGetValue(roomData.Id, out var room))
                    rooms.Add(room);
                else
                    rooms.Add(Room.Wrap(roomData));

            }

            return rooms;
        }

        /// <summary>
        /// Sort rooms suitable for navigator display
        /// </summary>
        public static List<Room> SortRooms(List<Room> list)
        {
            return list.OrderByDescending(x => x.Data.Rating)
                        .ThenByDescending(x => x.Data.UsersNow).ToList();
        }

        #endregion
    }
}

File: euclid/src/main/java/euclid/game/room/RoomManager.java
package euclid.game.room;

import euclid.game.ILoadable;
import euclid.storage.database.access.RoomDao;
import euclid.storage.database.data.RoomData;
import euclid.storage.database.data.RoomModelData;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

public class RoomManager implements ILoadable {

    public static final RoomManager INSTANCE = new RoomManager();

    private final ConcurrentHashMap<Integer, Room> rooms;
    private List<RoomModel> roomModels;

    private RoomManager() {
        this.rooms = new ConcurrentHashMap<>();
    }

    public static RoomManager getInstance() {
        return INSTANCE;
    }

    @Override
    public void load() {
        List<RoomModelData> models = RoomDao.getModels();
        roomModels = new ArrayList<>();
        for (RoomModelData modelData : models) {
            roomModels.add(new RoomModel(modelData));
        }
    }

    public ConcurrentHashMap<Integer, Room> getRooms() {
        return rooms;
    }

    public List<RoomModel> getRoomModels() {
        return roomModels;
    }

    public boolean hasRoom(int roomId) {
        return rooms.containsKey(roomId);
    }

    public void removeRoom(int roomId) {
        rooms.remove(roomId);
    }

    public void addRoom(Room room) {
        if (room == null)
            return;
        if (rooms.containsKey(room.getData().getId()))
            return;
        rooms.putIfAbsent(room.getData().getId(), room);
    }

    public Room getRoom(int roomId) {
        Room room = rooms.get(roomId);
        if (room != null)
            return room;

        RoomData data = RoomDao.getRoomData(roomId);
        if (data != null) {
            return new Room(data);
        }

        return null;
    }

    public List<Room> replaceQueryRooms(List<RoomData> roomsFromDatabase) {
        List<Room> rooms = new ArrayList<>();

        for (RoomData roomData : roomsFromDatabase) {
            Room room = this.rooms.get(roomData.getId());
            if (room != null)
                rooms.add(room);
            else
                rooms.add(Room.wrap(roomData));
        }

        return rooms;
    }

    public static List<Room> sortRooms(List<Room> list) {
        return list.stream()
                .sorted(Comparator.comparingInt((Room r) -> r.getData().getRating()).reversed()
                        .thenComparing(Comparator.comparingInt((Room r) -> r.getData().getUsersNow()).reversed()))
                .toList();
    }
}
